// carbon-components-react/cjs/components/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Icon = require('../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  small: _propTypes2.default.bool,
  kind: _propTypes2.default.oneOf(['primary', 'secondary', 'danger', 'ghost']).isRequired,
  href: _propTypes2.default.string,
  tabIndex: _propTypes2.default.number,
  type: _propTypes2.default.oneOf(['button', 'reset', 'submit']),
  role: _propTypes2.default.string,
  icon: _propTypes2.default.string,
  iconDescription: function iconDescription(props) {
    if (props.icon && !props.iconDescription) {
      return new Error('icon property specified without also providing an iconDescription property.');
    }
    return undefined;
  }
};

var defaultProps = {
  iconDescription: 'Provide icon description if icon is used',
  tabIndex: 0,
  type: 'button',
  disabled: false,
  small: false,
  kind: 'primary'
};

var Button = function Button(_ref) {
  var children = _ref.children,
      className = _ref.className,
      disabled = _ref.disabled,
      small = _ref.small,
      kind = _ref.kind,
      href = _ref.href,
      tabIndex = _ref.tabIndex,
      type = _ref.type,
      icon = _ref.icon,
      iconDescription = _ref.iconDescription,
      other = _objectWithoutProperties(_ref, ['children', 'className', 'disabled', 'small', 'kind', 'href', 'tabIndex', 'type', 'icon', 'iconDescription']);

  var buttonClasses = (0, _classnames2.default)(className, {
    'bx--btn': true,
    'bx--btn--sm': small,
    'bx--btn--primary': kind === 'primary',
    'bx--btn--danger': kind === 'danger',
    'bx--btn--secondary': kind === 'secondary',
    'bx--btn--ghost': kind === 'ghost'
  });

  var commonProps = {
    tabIndex: tabIndex,
    className: buttonClasses
  };

  var buttonImage = icon ? _react2.default.createElement(_Icon2.default, { name: icon, description: iconDescription, className: 'bx--btn__icon' }) : null;

  var button = _react2.default.createElement(
    'button',
    _extends({}, other, commonProps, { disabled: disabled, type: type }),
    children,
    buttonImage
  );

  var anchor = _react2.default.createElement(
    'a',
    _extends({}, other, commonProps, { href: href, role: 'button' }),
    children,
    buttonImage
  );

  return href ? anchor : button;
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

exports.default = Button;
