// carbon-components-react/cjs/components/DropdownItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  value: _propTypes2.default.string.isRequired,
  itemText: _propTypes2.default.string.isRequired,
  className: _propTypes2.default.string,
  onClick: _propTypes2.default.func
};

var defaultProps = {
  onClick: /* istanbul ignore next */function onClick() {}
};

var DropdownItem = function DropdownItem(_ref) {
  var className = _ref.className,
      value = _ref.value,
      itemText = _ref.itemText,
      onClick = _ref.onClick,
      other = _objectWithoutProperties(_ref, ['className', 'value', 'itemText', 'onClick']);

  var dropdownItemClasses = (0, _classnames2.default)(_defineProperty({
    'bx--dropdown-item': true
  }, className, className));

  var handleClick = function handleClick() {
    var info = {
      value: value,
      itemText: itemText
    };
    onClick(info);
  };

  return _react2.default.createElement(
    'li',
    _extends({}, other, {
      value: value,
      className: dropdownItemClasses,
      onClick: handleClick
    }),
    _react2.default.createElement(
      'a',
      {
        href: '#',
        onClick: /* istanbul ignore next */function onClick(evt) {
          return evt.preventDefault();
        },
        className: 'bx--dropdown-link'
      },
      itemText
    )
  );
};

DropdownItem.propTypes = propTypes;
DropdownItem.defaultProps = defaultProps;

exports.default = DropdownItem;
