// carbon-components-react/cjs/components/Footer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Link = require('./Link');

var _Link2 = _interopRequireDefault(_Link);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  labelOne: _propTypes2.default.string,
  linkTextOne: _propTypes2.default.string,
  linkHrefOne: _propTypes2.default.string,
  labelTwo: _propTypes2.default.string,
  linkTextTwo: _propTypes2.default.string,
  linkHrefTwo: _propTypes2.default.string,
  buttonText: _propTypes2.default.string
};

var defaultProps = {
  labelOne: 'Need Help?',
  linkTextOne: 'Contact Bluemix Sales',
  linkHrefOne: '#',
  labelTwo: 'Estimate Monthly Cost',
  linkTextTwo: 'Cost Calculator',
  linkHrefTwo: '#',
  buttonText: 'Create'
};

var Footer = function Footer(_ref) {
  var className = _ref.className,
      children = _ref.children,
      labelOne = _ref.labelOne,
      linkTextOne = _ref.linkTextOne,
      linkHrefOne = _ref.linkHrefOne,
      labelTwo = _ref.labelTwo,
      linkTextTwo = _ref.linkTextTwo,
      linkHrefTwo = _ref.linkHrefTwo,
      buttonText = _ref.buttonText,
      other = _objectWithoutProperties(_ref, ['className', 'children', 'labelOne', 'linkTextOne', 'linkHrefOne', 'labelTwo', 'linkTextTwo', 'linkHrefTwo', 'buttonText']);

  var classNames = (0, _classnames2.default)('bx--footer bx--footer--bottom-fixed', className);

  var footer = children ? _react2.default.createElement(
    'footer',
    _extends({}, other, { className: classNames }),
    children
  ) : _react2.default.createElement(
    'footer',
    _extends({}, other, { className: classNames }),
    _react2.default.createElement(
      'div',
      { className: 'bx--footer-info' },
      _react2.default.createElement(
        'div',
        { className: 'bx--footer-info__item' },
        _react2.default.createElement(
          'p',
          { className: 'bx--footer-label' },
          labelOne
        ),
        _react2.default.createElement(
          _Link2.default,
          { href: linkHrefOne },
          linkTextOne
        )
      ),
      _react2.default.createElement(
        'div',
        { className: 'bx--footer-info__item' },
        _react2.default.createElement(
          'p',
          { className: 'bx--footer-label' },
          labelTwo
        ),
        _react2.default.createElement(
          _Link2.default,
          { href: linkHrefTwo },
          linkTextTwo
        )
      )
    ),
    _react2.default.createElement(
      'div',
      { className: 'bx--footer-cta' },
      _react2.default.createElement(
        _Button2.default,
        { type: 'submit' },
        buttonText
      )
    )
  );

  return footer;
};

Footer.propTypes = propTypes;
Footer.defaultProps = defaultProps;

exports.default = Footer;
