// carbon-components-react/cjs/components/OverflowMenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  className: _propTypes2.default.string,
  itemText: _propTypes2.default.string.isRequired,
  hasDivider: _propTypes2.default.bool,
  isDelete: _propTypes2.default.bool,
  onBlur: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  onFocus: _propTypes2.default.func,
  onKeyDown: _propTypes2.default.func,
  onKeyUp: _propTypes2.default.func,
  onMouseDown: _propTypes2.default.func,
  onMouseEnter: _propTypes2.default.func,
  onMouseLeave: _propTypes2.default.func,
  onMouseUp: _propTypes2.default.func,
  closeMenu: _propTypes2.default.func
};

var defaultProps = {
  hasDivider: false,
  isDelete: false,
  itemText: 'Provide itemText',
  onClick: function onClick() {}
};

var OverflowMenuItem = function OverflowMenuItem(_ref) {
  var className = _ref.className,
      itemText = _ref.itemText,
      hasDivider = _ref.hasDivider,
      isDelete = _ref.isDelete,
      closeMenu = _ref.closeMenu,
      onClick = _ref.onClick,
      other = _objectWithoutProperties(_ref, ['className', 'itemText', 'hasDivider', 'isDelete', 'closeMenu', 'onClick']);

  var overflowMenuBtnClasses = (0, _classnames2.default)([className], 'bx--overflow-menu-options__btn');

  var overflowMenuItemClasses = (0, _classnames2.default)('bx--overflow-menu-options__option', {
    'bx--overflow-menu--divider': hasDivider,
    'bx--overflow-menu-options__option--danger': isDelete
  });

  var handleClick = function handleClick(evt) {
    onClick(evt);
    closeMenu();
  };

  var item = _react2.default.createElement(
    'li',
    { className: overflowMenuItemClasses },
    _react2.default.createElement(
      'button',
      _extends({}, other, {
        type: 'button',
        className: overflowMenuBtnClasses,
        onClick: handleClick
      }),
      itemText
    )
  );

  return item;
};

OverflowMenuItem.propTypes = propTypes;
OverflowMenuItem.defaultProps = defaultProps;

exports.default = OverflowMenuItem;
