// carbon-components-react/cjs/components/TableData.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  iconClassName: _propTypes2.default.string,
  expanded: _propTypes2.default.bool
};

var TableData = function TableData(props) {
  var children = props.children,
      className = props.className,
      iconClassName = props.iconClassName,
      expanded = props.expanded,
      other = _objectWithoutProperties(props, ['children', 'className', 'iconClassName', 'expanded']);

  var tableDataClasses = (0, _classnames2.default)(className);

  var iconClasses = (0, _classnames2.default)(iconClassName, 'bx--table-expand__svg');

  var style = expanded ? {
    transform: 'rotate(90deg)'
  } : {};

  return _react2.default.createElement(
    'td',
    _extends({}, other, { className: tableDataClasses }),
    expanded === undefined ? children : _react2.default.createElement(_Icon2.default, {
      className: iconClasses,
      name: 'chevron--right',
      description: 'expand row',
      style: style
    })
  );
};

TableData.propTypes = propTypes;

exports.default = TableData;
