// carbon-components-react/cjs/components/Table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  containerClassName: _propTypes2.default.string
};

var Table = function Table(props) {
  var children = props.children,
      className = props.className,
      containerClassName = props.containerClassName,
      other = _objectWithoutProperties(props, ['children', 'className', 'containerClassName']);

  var tableClasses = (0, _classnames2.default)(className, 'bx--responsive-table');

  var tableContainerClasses = (0, _classnames2.default)(containerClassName, 'bx--responsive-table-container');

  return _react2.default.createElement(
    'div',
    { className: tableContainerClasses },
    _react2.default.createElement(
      'table',
      _extends({}, other, {
        className: tableClasses,
        style: {
          borderCollapse: 'collapse',
          borderSpacing: 0
        }
      }),
      children
    )
  );
};

Table.propTypes = propTypes;

exports.default = Table;
