// carbon-components-react/cjs/components/TextInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  className: _propTypes2.default.string,
  defaultValue: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  disabled: _propTypes2.default.bool,
  id: _propTypes2.default.string.isRequired,
  labelText: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  placeholder: _propTypes2.default.string,
  type: _propTypes2.default.string,
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  hideLabel: _propTypes2.default.bool,
  invalid: _propTypes2.default.bool,
  invalidText: _propTypes2.default.string
};

var defaultProps = {
  className: 'bx--text__input',
  disabled: false,
  type: 'text',
  onChange: function onChange() {},
  onClick: function onClick() {},
  invalid: false,
  labelText: 'Provide labelText',
  invalidText: 'Provide invalidText'
};

var TextInput = function TextInput(_ref) {
  var labelText = _ref.labelText,
      className = _ref.className,
      id = _ref.id,
      placeholder = _ref.placeholder,
      type = _ref.type,
      _onChange = _ref.onChange,
      _onClick = _ref.onClick,
      hideLabel = _ref.hideLabel,
      invalid = _ref.invalid,
      invalidText = _ref.invalidText,
      other = _objectWithoutProperties(_ref, ['labelText', 'className', 'id', 'placeholder', 'type', 'onChange', 'onClick', 'hideLabel', 'invalid', 'invalidText']);

  var textInputProps = {
    id: id,
    onChange: function onChange(evt) {
      if (!other.disabled) {
        _onChange(evt);
      }
    },
    onClick: function onClick(evt) {
      if (!other.disabled) {
        _onClick(evt);
      }
    },
    placeholder: placeholder,
    type: type
  };

  var textInputClasses = (0, _classnames2.default)('bx--text-input', className);
  var labelClasses = (0, _classnames2.default)('bx--label', {
    'bx--visually-hidden': hideLabel
  });

  var label = labelText ? _react2.default.createElement(
    'label',
    { htmlFor: id, className: labelClasses },
    labelText
  ) : null;

  var error = invalid ? _react2.default.createElement(
    'div',
    { className: 'bx--form-requirement' },
    invalidText
  ) : null;

  var input = invalid ? _react2.default.createElement('input', _extends({}, other, textInputProps, { 'data-invalid': true, className: textInputClasses })) : _react2.default.createElement('input', _extends({}, other, textInputProps, { className: textInputClasses }));

  return _react2.default.createElement(
    'div',
    { className: 'bx--form-item' },
    label,
    input,
    error
  );
};

TextInput.propTypes = propTypes;
TextInput.defaultProps = defaultProps;

exports.default = TextInput;
