// carbon-components-react/cjs/components/TextArea.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  className: _propTypes2.default.string,
  cols: _propTypes2.default.number,
  defaultValue: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  disabled: _propTypes2.default.bool,
  id: _propTypes2.default.string,
  labelText: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  placeholder: _propTypes2.default.string,
  rows: _propTypes2.default.number,
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  invalid: _propTypes2.default.bool,
  invalidText: _propTypes2.default.string
};

var defaultProps = {
  disabled: false,
  onChange: function onChange() {},
  onClick: function onClick() {},
  placeholder: 'Hint text here',
  rows: 4,
  cols: 50,
  invalid: false,
  labelText: 'Provide labelText',
  invalidText: 'Provide invalidText'
};

var Textarea = function Textarea(_ref) {
  var className = _ref.className,
      id = _ref.id,
      labelText = _ref.labelText,
      _onChange = _ref.onChange,
      _onClick = _ref.onClick,
      invalid = _ref.invalid,
      invalidText = _ref.invalidText,
      other = _objectWithoutProperties(_ref, ['className', 'id', 'labelText', 'onChange', 'onClick', 'invalid', 'invalidText']);

  var textareaProps = {
    id: id,
    onChange: function onChange(evt) {
      if (!other.disabled) {
        _onChange(evt);
      }
    },
    onClick: function onClick(evt) {
      if (!other.disabled) {
        _onClick(evt);
      }
    }
  };

  var textareaClasses = (0, _classnames2.default)('bx--text-area', className);
  var label = labelText ? _react2.default.createElement(
    'label',
    { htmlFor: id, className: 'bx--label' },
    labelText
  ) : null;

  var error = invalid ? _react2.default.createElement(
    'div',
    { className: 'bx--form-requirement' },
    invalidText
  ) : null;

  var input = invalid ? _react2.default.createElement('textarea', _extends({}, other, textareaProps, { className: textareaClasses, 'data-invalid': true })) : _react2.default.createElement('textarea', _extends({}, other, textareaProps, { className: textareaClasses }));

  return _react2.default.createElement(
    'div',
    { className: 'bx--form-item' },
    label,
    input,
    error
  );
};

Textarea.propTypes = propTypes;
Textarea.defaultProps = defaultProps;

exports.default = Textarea;
