// carbon-components-react/cjs/components/Tooltip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _FloatingMenu = require('../internal/FloatingMenu');

var _FloatingMenu2 = _interopRequireDefault(_FloatingMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Tooltip = function (_Component) {
  _inherits(Tooltip, _Component);

  function Tooltip() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Tooltip);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      open: _this.props.open
    }, _this.getTriggerPosition = function () {
      var triggerPosition = _this.triggerEl.getBoundingClientRect();
      _this.setState({ triggerPosition: triggerPosition });
    }, _this.handleMouse = function (direction) {
      if (direction === 'over') {
        _this.getTriggerPosition();
        _this.setState({ open: true });
      } else {
        _this.setState({ open: false });
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Tooltip, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      requestAnimationFrame(function () {
        _this2.getTriggerPosition();
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          children = _props.children,
          className = _props.className,
          direction = _props.direction,
          triggerText = _props.triggerText,
          showIcon = _props.showIcon,
          iconName = _props.iconName,
          iconDescription = _props.iconDescription,
          other = _objectWithoutProperties(_props, ['children', 'className', 'direction', 'triggerText', 'showIcon', 'iconName', 'iconDescription']);

      var tooltipClasses = (0, _classnames2.default)('bx--tooltip', { 'bx--tooltip--shown': this.state.open }, className);

      var menuOffset = { left: 5, top: 10 };

      return _react2.default.createElement(
        'div',
        null,
        showIcon ? _react2.default.createElement(
          'div',
          { className: 'bx--tooltip__trigger' },
          triggerText,
          _react2.default.createElement(
            'div',
            {
              ref: function ref(node) {
                _this3.triggerEl = node;
              },
              onMouseOver: function onMouseOver() {
                return _this3.handleMouse('over');
              },
              onMouseOut: function onMouseOut() {
                return _this3.handleMouse('out');
              },
              onFocus: function onFocus() {
                return _this3.handleMouse('over');
              },
              onBlur: function onBlur() {
                return _this3.handleMouse('out');
              }
            },
            _react2.default.createElement(_Icon2.default, { name: iconName, description: iconDescription, tabIndex: '0' })
          )
        ) : _react2.default.createElement(
          'div',
          {
            className: 'bx--tooltip__trigger',
            ref: function ref(node) {
              _this3.triggerEl = node;
            },
            onMouseOver: function onMouseOver() {
              return _this3.handleMouse('over');
            },
            onMouseOut: function onMouseOut() {
              return _this3.handleMouse('out');
            },
            onFocus: function onFocus() {
              return _this3.handleMouse('over');
            },
            onBlur: function onBlur() {
              return _this3.handleMouse('out');
            }
          },
          triggerText
        ),
        _react2.default.createElement(
          _FloatingMenu2.default,
          {
            menuPosition: this.state.triggerPosition,
            menuDirection: direction,
            menuOffset: menuOffset
          },
          _react2.default.createElement(
            'div',
            _extends({
              className: tooltipClasses
            }, other, {
              'data-floating-menu-direction': direction
            }),
            children
          )
        )
      );
    }
  }]);

  return Tooltip;
}(_react.Component);

Tooltip.propTypes = {
  open: _propTypes2.default.bool,
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  direction: _propTypes2.default.oneOf(['bottom', 'top']),
  triggerText: _propTypes2.default.string,
  showIcon: _propTypes2.default.bool,
  iconName: _propTypes2.default.string,
  iconDescription: _propTypes2.default.string
};
Tooltip.defaultProps = {
  open: false,
  direction: 'bottom',
  showIcon: true,
  iconName: 'info--glyph',
  iconDescription: 'tooltip',
  triggerText: 'Provide triggerText'
};
exports.default = Tooltip;
