// carbon-components-react/cjs/components/Toolbar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolbarDivider = exports.ToolbarOption = exports.ToolbarTitle = exports.ToolbarItem = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _ToolbarSearch = require('./ToolbarSearch');

var _ToolbarSearch2 = _interopRequireDefault(_ToolbarSearch);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string
};

var toolbarItemPropTypes = {
  children: _propTypes2.default.node,
  type: _propTypes2.default.string,
  placeHolderText: _propTypes2.default.string
};

var toolbarTitlePropTypes = {
  title: _propTypes2.default.string
};

var toolbarTitleDefaultProps = {
  title: _propTypes2.default.string
};

var toolbarItemDefaultProps = {
  placeHolderText: 'Provide placeHolderText'
};

var toolbarOptionPropTypes = {
  children: _propTypes2.default.node
};

var Toolbar = function Toolbar(_ref) {
  var children = _ref.children,
      className = _ref.className,
      other = _objectWithoutProperties(_ref, ['children', 'className']);

  var wrapperClasses = (0, _classnames2.default)('bx--toolbar', className);

  return _react2.default.createElement(
    'div',
    _extends({ className: wrapperClasses }, other),
    children
  );
};

var ToolbarItem = function ToolbarItem(_ref2) {
  var children = _ref2.children,
      type = _ref2.type,
      placeHolderText = _ref2.placeHolderText;

  var toolbarItem = type === 'search' ? _react2.default.createElement(_ToolbarSearch2.default, { placeHolderText: placeHolderText }) : children;
  return toolbarItem;
};

var ToolbarTitle = function ToolbarTitle(_ref3) {
  var title = _ref3.title;
  return _react2.default.createElement(
    'li',
    { className: 'bx--toolbar-menu__title' },
    title
  );
};

var ToolbarOption = function ToolbarOption(_ref4) {
  var children = _ref4.children;
  return _react2.default.createElement(
    'li',
    { className: 'bx--toolbar-menu__option' },
    children
  );
};

var ToolbarDivider = function ToolbarDivider() {
  return _react2.default.createElement('hr', { className: 'bx--toolbar-menu__divider' });
};

Toolbar.propTypes = propTypes;
ToolbarItem.propTypes = toolbarItemPropTypes;
ToolbarTitle.propTypes = toolbarTitlePropTypes;
ToolbarOption.propTypes = toolbarOptionPropTypes;

ToolbarItem.defaultProps = toolbarItemDefaultProps;
ToolbarTitle.defaultProps = toolbarTitleDefaultProps;

exports.default = Toolbar;
exports.ToolbarItem = ToolbarItem;
exports.ToolbarTitle = ToolbarTitle;
exports.ToolbarOption = ToolbarOption;
exports.ToolbarDivider = ToolbarDivider;
