// carbon-components-react/es/components/AccordionItem.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Icon from './Icon';

var AccordionItem = function (_Component) {
  _inherits(AccordionItem, _Component);

  function AccordionItem() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, AccordionItem);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = AccordionItem.__proto__ || Object.getPrototypeOf(AccordionItem)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      open: _this.props.open
    }, _this.handleClick = function (evt) {
      _this.props.onClick(evt);
    }, _this.handleHeadingClick = function (evt) {
      var open = !_this.state.open;
      _this.setState({ open: open });
      _this.props.onHeadingClick({ isOpen: open, event: evt });
    }, _this.handleKeyPress = function (evt) {
      var isKeyPressTarget = evt.target === evt.currentTarget;
      var isValidKeyPress = [13, 32].indexOf(evt.which) !== -1;

      if (isKeyPressTarget && isValidKeyPress) {
        _this.handleHeadingClick(evt);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(AccordionItem, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(_ref2) {
      var open = _ref2.open;

      if (open !== this.props.open) {
        this.setState({ open: open });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          title = _props.title,
          children = _props.children,
          onClick = _props.onClick,
          onHeadingClick = _props.onHeadingClick,
          other = _objectWithoutProperties(_props, ['className', 'title', 'children', 'onClick', 'onHeadingClick']);

      var classNames = classnames({
        'bx--accordion__item--active': this.state.open
      }, 'bx--accordion__item', className);
      return React.createElement(
        'li',
        _extends({
          className: classNames,
          onClick: this.handleClick,
          onKeyPress: this.handleKeyPress
        }, other, {
          tabIndex: '0'
        }),
        React.createElement(
          'div',
          { className: 'bx--accordion__heading', onClick: this.handleHeadingClick },
          React.createElement(Icon, {
            className: 'bx--accordion__arrow',
            name: 'chevron--right',
            description: 'Expand/Collapse'
          }),
          React.createElement(
            'p',
            { className: 'bx--accordion__title' },
            title
          )
        ),
        React.createElement(
          'div',
          { className: 'bx--accordion__content' },
          children
        )
      );
    }
  }]);

  return AccordionItem;
}(Component);

AccordionItem.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  title: PropTypes.string,
  open: PropTypes.bool,
  onClick: PropTypes.func,
  onHeadingClick: PropTypes.func
};
AccordionItem.defaultProps = {
  title: 'title',
  open: false,
  onClick: function onClick() {},
  onHeadingClick: function onHeadingClick() {}
};


export default AccordionItem;
