// carbon-components-react/es/components/Checkbox.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import Icon from './Icon';

var propTypes = {
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  labelText: PropTypes.string,
  onChange: PropTypes.func,
  iconDescription: PropTypes.string
};

var defaultProps = {
  onChange: function onChange() {},
  labelText: 'Provide checkbox label text',
  iconDescription: 'Provide icon description for a11y'
};

var Checkbox = function Checkbox(_ref) {
  var className = _ref.className,
      id = _ref.id,
      labelText = _ref.labelText,
      _onChange = _ref.onChange,
      iconDescription = _ref.iconDescription,
      other = _objectWithoutProperties(_ref, ['className', 'id', 'labelText', 'onChange', 'iconDescription']);

  var input = void 0;
  var wrapperClasses = classNames('bx--checkbox-label', className);

  return React.createElement(
    'div',
    { className: 'bx--form-item bx--checkbox-wrapper' },
    React.createElement(
      'label',
      { htmlFor: id, className: wrapperClasses },
      React.createElement('input', _extends({}, other, {
        type: 'checkbox',
        onChange: function onChange(evt) {
          _onChange(input.checked, id, evt);
        },
        className: 'bx--checkbox',
        id: id,
        ref: function ref(el) {
          input = el;
        }
      })),
      React.createElement(
        'span',
        { className: 'bx--checkbox-appearance' },
        React.createElement(Icon, {
          className: 'bx--checkbox-checkmark',
          description: iconDescription,
          name: 'checkmark'
        })
      ),
      React.createElement(
        'span',
        { className: 'bx--checkbox-label-text' },
        labelText
      )
    )
  );
};

Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;

export default Checkbox;
