// carbon-components-react/es/components/InteriorLeftNavItem.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';

var propTypes = {
  className: PropTypes.string,
  href: PropTypes.string.isRequired,
  activeHref: PropTypes.string,
  tabIndex: PropTypes.number,
  onClick: PropTypes.func,
  blankTarget: PropTypes.bool,
  children: PropTypes.node,
  label: PropTypes.string.isRequired
};

var defaultProps = {
  activeHref: '#',
  tabIndex: 0,
  label: 'InteriorLeftNavItem Label',
  onClick: /* istanbul ignore next */function onClick() {}
};

var newChild = function newChild(children, tabIndex) {
  var child = React.Children.only(children);
  return React.cloneElement(React.Children.only(child), {
    tabIndex: tabIndex,
    className: 'left-nav-list__item-link'
  });
};

var InteriorLeftNavItem = function InteriorLeftNavItem(_ref) {
  var className = _ref.className,
      href = _ref.href,
      activeHref = _ref.activeHref,
      _onClick = _ref.onClick,
      tabIndex = _ref.tabIndex,
      children = _ref.children,
      label = _ref.label,
      other = _objectWithoutProperties(_ref, ['className', 'href', 'activeHref', 'onClick', 'tabIndex', 'children', 'label']);

  var classNames = classnames('left-nav-list__item', className, {
    'left-nav-list__item--active': activeHref === href
  });

  return React.createElement(
    'li',
    _extends({
      role: 'menuitem',
      tabIndex: tabIndex,
      className: classNames,
      onClick: function onClick(evt) {
        return _onClick(evt, href);
      },
      onKeyPress: function onKeyPress(evt) {
        return _onClick(evt, href);
      }
    }, other),
    children ? newChild(children, tabIndex) : React.createElement(
      'a',
      { className: 'left-nav-list__item-link', href: href, tabIndex: tabIndex },
      label
    )
  );
};

InteriorLeftNavItem.propTypes = propTypes;
InteriorLeftNavItem.defaultProps = defaultProps;

export default InteriorLeftNavItem;
