// carbon-components-react/es/components/Notification.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import Icon from './Icon';

var Notification = function (_Component) {
  _inherits(Notification, _Component);

  function Notification() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Notification);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Notification.__proto__ || Object.getPrototypeOf(Notification)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      open: true
    }, _this.handleCloseButtonClick = function (evt) {
      _this.setState({ open: false });
      _this.props.onCloseButtonClick(evt);
    }, _this.useIconName = function (kindProp) {
      var isSuccess = kindProp === 'success';
      return isSuccess ? 'checkmark--glyph' : kindProp + '--glyph';
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Notification, [{
    key: 'render',
    value: function render() {
      if (!this.state.open) {
        return null;
      }

      var _props = this.props,
          onCloseButtonClick = _props.onCloseButtonClick,
          iconDescription = _props.iconDescription,
          className = _props.className,
          caption = _props.caption,
          subtitle = _props.subtitle,
          title = _props.title,
          kind = _props.kind,
          other = _objectWithoutProperties(_props, ['onCloseButtonClick', 'iconDescription', 'className', 'caption', 'subtitle', 'title', 'kind']);

      var notificationClasses = {
        toast: classNames('bx--toast-notification', _defineProperty({}, 'bx--toast-notification--' + this.props.kind, this.props.kind), className),
        inline: classNames('bx--inline-notification', _defineProperty({}, 'bx--inline-notification--' + this.props.kind, this.props.kind), className)
      };

      var commonProps = {
        alert: {
          role: 'alert',
          kind: kind
        },
        button: {
          type: 'button',
          onClick: this.handleCloseButtonClick
        }
      };

      var toastHTML = React.createElement(
        'div',
        _extends({}, other, commonProps.alert, { className: notificationClasses.toast }),
        React.createElement(
          'div',
          { className: 'bx--toast-notification__details' },
          React.createElement(
            'h3',
            { className: 'bx--toast-notification__title' },
            title
          ),
          React.createElement(
            'p',
            { className: 'bx--toast-notification__subtitle' },
            subtitle
          ),
          React.createElement(
            'p',
            { className: 'bx--toast-notification__caption' },
            caption
          )
        ),
        React.createElement(
          'button',
          _extends({}, commonProps.button, { className: 'bx--toast-notification__close-button' }),
          React.createElement(Icon, {
            description: this.props.iconDescription,
            className: 'bx--toast-notification__icon',
            'aria-label': 'close',
            name: 'close'
          })
        )
      );

      var inlineHTML = React.createElement(
        'div',
        _extends({}, other, commonProps.alert, { className: notificationClasses.inline }),
        React.createElement(
          'div',
          { className: 'bx--inline-notification__details' },
          React.createElement(Icon, {
            description: this.props.iconDescription,
            className: 'bx--inline-notification__icon',
            'aria-label': 'close',
            name: this.useIconName(kind)
          }),
          React.createElement(
            'div',
            { className: 'bx--inline-notification__text-wrapper' },
            React.createElement(
              'p',
              { className: 'bx--inline-notification__title' },
              title
            ),
            React.createElement(
              'p',
              { className: 'bx--inline-notification__subtitle' },
              subtitle
            )
          )
        ),
        React.createElement(
          'button',
          _extends({}, commonProps.button, { className: 'bx--inline-notification__close-button' }),
          React.createElement(Icon, {
            description: this.props.iconDescription,
            className: 'bx--inline-notification__close-icon',
            'aria-label': 'close',
            name: 'close'
          })
        )
      );

      return caption ? toastHTML : inlineHTML;
    }
  }]);

  return Notification;
}(Component);

Notification.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  kind: PropTypes.oneOf(['error', 'info', 'success', 'warning']).isRequired,
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
  caption: PropTypes.string,
  onCloseButtonClick: PropTypes.func,
  iconDescription: PropTypes.string.isRequired
};
Notification.defaultProps = {
  onCloseButtonClick: function onCloseButtonClick() {},
  iconDescription: 'closes notification',
  title: 'Provide a title',
  subtitle: 'Provide a subtitle'
};


export default Notification;
