// carbon-components-react/es/components/Pagination.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Icon from './Icon';
import Select from './Select';
import SelectItem from './SelectItem';
import TextInput from './TextInput';
import { equals } from '../lib/array';

var Pagination = function (_Component) {
  _inherits(Pagination, _Component);

  function Pagination() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Pagination);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Pagination.__proto__ || Object.getPrototypeOf(Pagination)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      page: _this.props.page,
      pageSize: _this.props.pageSize && _this.props.pageSizes.includes(_this.props.pageSize) ? _this.props.pageSize : _this.props.pageSizes[0]
    }, _this.handleSizeChange = function (evt) {
      var pageSize = Number(evt.target.value);
      _this.setState({ pageSize: pageSize, page: 1 });
      _this.props.onChange({ page: 1, pageSize: pageSize });
    }, _this.handlePageInputChange = function (evt) {
      var page = Number(evt.target.value);
      if (page > 0 && page <= Math.ceil(_this.props.totalItems / _this.state.pageSize)) {
        _this.setState({ page: page });
        _this.props.onChange({ page: page, pageSize: _this.state.pageSize });
      }
    }, _this.incrementPage = function () {
      var page = _this.state.page + 1;
      _this.setState({ page: page });
      _this.props.onChange({ page: page, pageSize: _this.state.pageSize });
    }, _this.decrementPage = function () {
      var page = _this.state.page - 1;
      _this.setState({ page: page });
      _this.props.onChange({ page: page, pageSize: _this.state.pageSize });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Pagination, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.uniqueId = '' + Math.floor(Math.random() * 0xffff);
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(_ref2) {
      var pageSizes = _ref2.pageSizes,
          page = _ref2.page,
          pageSize = _ref2.pageSize;

      if (!equals(pageSizes, this.props.pageSizes)) {
        this.setState({ pageSize: pageSizes[0], page: 1 });
      }
      if (page !== this.props.page) {
        this.setState({
          page: page
        });
      }
      if (pageSize !== this.props.pageSize) {
        this.setState({ pageSize: pageSize });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          backwardText = _props.backwardText,
          className = _props.className,
          forwardText = _props.forwardText,
          id = _props.id,
          itemsPerPageText = _props.itemsPerPageText,
          itemRangeText = _props.itemRangeText,
          pageNumberText = _props.pageNumberText,
          pageRangeText = _props.pageRangeText,
          pageSize = _props.pageSize,
          pageSizes = _props.pageSizes,
          itemText = _props.itemText,
          pageText = _props.pageText,
          pagesUnknown = _props.pagesUnknown,
          isLastPage = _props.isLastPage,
          pageInputDisabled = _props.pageInputDisabled,
          totalItems = _props.totalItems,
          onChange = _props.onChange,
          pageNumber = _props.page,
          other = _objectWithoutProperties(_props, ['backwardText', 'className', 'forwardText', 'id', 'itemsPerPageText', 'itemRangeText', 'pageNumberText', 'pageRangeText', 'pageSize', 'pageSizes', 'itemText', 'pageText', 'pagesUnknown', 'isLastPage', 'pageInputDisabled', 'totalItems', 'onChange', 'page']);

      var statePage = this.state.page;
      var statePageSize = this.state.pageSize;
      var classNames = classnames('bx--pagination', className);
      var inputId = id || this.uniqueId;

      return React.createElement(
        'div',
        _extends({ className: classNames }, other),
        React.createElement(
          'div',
          { className: 'bx--pagination__left' },
          React.createElement(
            Select,
            {
              id: 'bx-pagination-select-' + inputId,
              labelText: itemsPerPageText,
              hideLabel: true,
              onChange: this.handleSizeChange,
              value: statePageSize
            },
            pageSizes.map(function (size) {
              return React.createElement(SelectItem, { key: size, value: size, text: String(size) });
            })
          ),
          React.createElement(
            'span',
            { className: 'bx--pagination__text' },
            itemsPerPageText,
            '\xA0\xA0|\xA0\xA0'
          ),
          React.createElement(
            'span',
            { className: 'bx--pagination__text' },
            pagesUnknown ? itemText(statePageSize * (statePage - 1) + 1, statePage * statePageSize) : itemRangeText(statePageSize * (statePage - 1) + 1, Math.min(statePage * statePageSize, totalItems), totalItems)
          )
        ),
        React.createElement(
          'div',
          { className: 'bx--pagination__right' },
          React.createElement(
            'span',
            { className: 'bx--pagination__text' },
            pagesUnknown ? pageText(statePage) : pageRangeText(statePage, Math.ceil(totalItems / statePageSize))
          ),
          React.createElement(
            'button',
            {
              className: 'bx--pagination__button bx--pagination__button--backward',
              onClick: this.decrementPage,
              disabled: this.props.disabled || statePage === 1
            },
            React.createElement(Icon, {
              className: 'bx--pagination__button-icon',
              name: 'chevron--left',
              description: backwardText
            })
          ),
          pageInputDisabled ? React.createElement(
            'span',
            { className: 'bx--pagination__text' },
            '|'
          ) : React.createElement(TextInput, {
            id: 'bx-pagination-input-' + this.id,
            placeholder: '0',
            value: statePage,
            onChange: this.handlePageInputChange,
            labelText: pageNumberText,
            hideLabel: true
          }),
          React.createElement(
            'button',
            {
              className: 'bx--pagination__button bx--pagination__button--forward',
              onClick: this.incrementPage,
              disabled: this.props.disabled || statePage === Math.ceil(totalItems / statePageSize) || isLastPage
            },
            React.createElement(Icon, {
              className: 'bx--pagination__button-icon',
              name: 'chevron--right',
              description: forwardText
            })
          )
        )
      );
    }
  }]);

  return Pagination;
}(Component);

Pagination.propTypes = {
  backwardText: PropTypes.string,
  className: PropTypes.string,
  itemRangeText: PropTypes.func,
  forwardText: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  itemsPerPageText: PropTypes.string,
  itemText: PropTypes.func,
  onChange: PropTypes.func,
  pageNumberText: PropTypes.string,
  pageRangeText: PropTypes.func,
  pageText: PropTypes.func,
  pageSizes: PropTypes.arrayOf(PropTypes.number).isRequired,
  totalItems: PropTypes.number,
  disabled: PropTypes.bool,
  page: PropTypes.number,
  pageSize: PropTypes.number,
  pagesUnknown: PropTypes.bool,
  isLastPage: PropTypes.bool,
  pageInputDisabled: PropTypes.bool
};
Pagination.defaultProps = {
  backwardText: 'Backward',
  itemRangeText: function itemRangeText(min, max, total) {
    return min + '-' + max + ' of ' + total + ' items';
  },
  forwardText: 'Forward',
  itemsPerPageText: 'items per page',
  onChange: function onChange() {},
  pageNumberText: 'Page Number',
  pageRangeText: function pageRangeText(current, total) {
    return current + ' of ' + total + ' pages';
  },
  disabled: false,
  page: 1,
  pagesUnknown: false,
  isLastPage: false,
  pageInputDisabled: false,
  itemText: function itemText(min, max) {
    return min + '-' + max + ' items';
  },
  pageText: function pageText(page) {
    return 'page ' + page;
  }
};


export default Pagination;
