// carbon-components-react/es/components/StructuredList.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import uid from '../lib/uniqueId';

var StructuredListWrapper = function (_Component) {
  _inherits(StructuredListWrapper, _Component);

  function StructuredListWrapper() {
    _classCallCheck(this, StructuredListWrapper);

    return _possibleConstructorReturn(this, (StructuredListWrapper.__proto__ || Object.getPrototypeOf(StructuredListWrapper)).apply(this, arguments));
  }

  _createClass(StructuredListWrapper, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          selection = _props.selection,
          className = _props.className,
          border = _props.border,
          other = _objectWithoutProperties(_props, ['children', 'selection', 'className', 'border']);

      var classes = classNames('bx--structured-list', className, {
        'bx--structured-list--border': border,
        'bx--structured-list--selection': selection
      });

      return React.createElement(
        'section',
        _extends({ className: classes }, other),
        children
      );
    }
  }]);

  return StructuredListWrapper;
}(Component);

StructuredListWrapper.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  border: PropTypes.bool,
  selection: PropTypes.bool
};
StructuredListWrapper.defaultProps = {
  border: false,
  selection: false
};

var StructuredListHead = function (_Component2) {
  _inherits(StructuredListHead, _Component2);

  function StructuredListHead() {
    _classCallCheck(this, StructuredListHead);

    return _possibleConstructorReturn(this, (StructuredListHead.__proto__ || Object.getPrototypeOf(StructuredListHead)).apply(this, arguments));
  }

  _createClass(StructuredListHead, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          children = _props2.children,
          className = _props2.className,
          other = _objectWithoutProperties(_props2, ['children', 'className']);

      var classes = classNames('bx--structured-list-thead', className);
      return React.createElement(
        'div',
        _extends({ className: classes }, other),
        children
      );
    }
  }]);

  return StructuredListHead;
}(Component);

StructuredListHead.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string
};

var StructuredListInput = function (_Component3) {
  _inherits(StructuredListInput, _Component3);

  function StructuredListInput() {
    _classCallCheck(this, StructuredListInput);

    return _possibleConstructorReturn(this, (StructuredListInput.__proto__ || Object.getPrototypeOf(StructuredListInput)).apply(this, arguments));
  }

  _createClass(StructuredListInput, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.uid = this.props.id || uid();
    }
  }, {
    key: 'render',
    value: function render() {
      var _props3 = this.props,
          className = _props3.className,
          value = _props3.value,
          name = _props3.name,
          title = _props3.title,
          other = _objectWithoutProperties(_props3, ['className', 'value', 'name', 'title']);

      var classes = classNames('bx--structured-list-input', className);
      return React.createElement('input', _extends({}, other, {
        type: 'radio',
        tabIndex: -1,
        id: this.uid,
        className: classes,
        value: value,
        name: name,
        title: title
      }));
    }
  }]);

  return StructuredListInput;
}(Component);

StructuredListInput.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  name: PropTypes.string,
  title: PropTypes.string,
  defaultChecked: PropTypes.bool,
  onChange: PropTypes.func
};
StructuredListInput.defaultProps = {
  onChange: function onChange() {},
  value: 'value',
  title: 'title'
};

var StructuredListRow = function (_Component4) {
  _inherits(StructuredListRow, _Component4);

  function StructuredListRow() {
    _classCallCheck(this, StructuredListRow);

    return _possibleConstructorReturn(this, (StructuredListRow.__proto__ || Object.getPrototypeOf(StructuredListRow)).apply(this, arguments));
  }

  _createClass(StructuredListRow, [{
    key: 'render',
    value: function render() {
      var _props4 = this.props,
          onKeyDown = _props4.onKeyDown,
          tabIndex = _props4.tabIndex,
          htmlFor = _props4.htmlFor,
          children = _props4.children,
          className = _props4.className,
          head = _props4.head,
          label = _props4.label,
          other = _objectWithoutProperties(_props4, ['onKeyDown', 'tabIndex', 'htmlFor', 'children', 'className', 'head', 'label']);

      var classes = classNames('bx--structured-list-row', className, {
        'bx--structured-list-row--header-row': head
      });

      return label ? React.createElement(
        'label',
        _extends({}, other, {
          tabIndex: tabIndex,
          className: classes,
          htmlFor: htmlFor,
          onKeyDown: onKeyDown
        }),
        children
      ) : React.createElement(
        'div',
        _extends({}, other, { className: classes }),
        children
      );
    }
  }]);

  return StructuredListRow;
}(Component);

StructuredListRow.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  head: PropTypes.bool,
  label: PropTypes.bool,
  htmlFor: PropTypes.string,
  tabIndex: PropTypes.number,
  onKeyDown: PropTypes.func
};
StructuredListRow.defaultProps = {
  htmlFor: 'unique id',
  head: false,
  label: false,
  tabIndex: 0,
  onKeyDown: function onKeyDown() {}
};

var StructuredListBody = function (_Component5) {
  _inherits(StructuredListBody, _Component5);

  function StructuredListBody() {
    var _ref;

    var _temp, _this5, _ret;

    _classCallCheck(this, StructuredListBody);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this5 = _possibleConstructorReturn(this, (_ref = StructuredListBody.__proto__ || Object.getPrototypeOf(StructuredListBody)).call.apply(_ref, [this].concat(args))), _this5), _this5.state = {
      labelRows: null,
      rowSelected: 0
    }, _temp), _possibleConstructorReturn(_this5, _ret);
  }

  _createClass(StructuredListBody, [{
    key: 'render',
    value: function render() {
      var _props5 = this.props,
          children = _props5.children,
          className = _props5.className,
          other = _objectWithoutProperties(_props5, ['children', 'className']);

      var classes = classNames('bx--structured-list-tbody', className);
      return React.createElement(
        'div',
        _extends({ className: classes }, other),
        children
      );
    }
  }]);

  return StructuredListBody;
}(Component);

StructuredListBody.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  head: PropTypes.bool,
  onKeyDown: PropTypes.func
};
StructuredListBody.defaultProps = {
  onKeyDown: function onKeyDown() {}
};

var StructuredListCell = function (_Component6) {
  _inherits(StructuredListCell, _Component6);

  function StructuredListCell() {
    _classCallCheck(this, StructuredListCell);

    return _possibleConstructorReturn(this, (StructuredListCell.__proto__ || Object.getPrototypeOf(StructuredListCell)).apply(this, arguments));
  }

  _createClass(StructuredListCell, [{
    key: 'render',
    value: function render() {
      var _props6 = this.props,
          children = _props6.children,
          className = _props6.className,
          head = _props6.head,
          noWrap = _props6.noWrap,
          other = _objectWithoutProperties(_props6, ['children', 'className', 'head', 'noWrap']);

      var classes = classNames(className, {
        'bx--structured-list-th': head,
        'bx--structured-list-td': !head,
        'bx--structured-list-content--nowrap': noWrap
      });

      return React.createElement(
        'div',
        _extends({ className: classes }, other),
        children
      );
    }
  }]);

  return StructuredListCell;
}(Component);

StructuredListCell.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  head: PropTypes.bool,
  noWrap: PropTypes.bool
};
StructuredListCell.defaultProps = {
  head: false,
  noWrap: false
};


export { StructuredListWrapper, StructuredListHead, StructuredListBody, StructuredListRow, StructuredListInput, StructuredListCell };
