// carbon-components/umd/components/detail-page-header/detail-page-header.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'lodash.debounce', '../../globals/js/misc/mixin', '../../globals/js/mixins/create-component', '../../globals/js/mixins/init-component-by-search', '../../globals/js/misc/on'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('lodash.debounce'), require('../../globals/js/misc/mixin'), require('../../globals/js/mixins/create-component'), require('../../globals/js/mixins/init-component-by-search'), require('../../globals/js/misc/on'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.lodash, global.mixin, global.createComponent, global.initComponentBySearch, global.on);
    global.detailPageHeader = mod.exports;
  }
})(this, function (exports, _lodash, _mixin2, _createComponent, _initComponentBySearch, _on) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _lodash2 = _interopRequireDefault(_lodash);

  var _mixin3 = _interopRequireDefault(_mixin2);

  var _createComponent2 = _interopRequireDefault(_createComponent);

  var _initComponentBySearch2 = _interopRequireDefault(_initComponentBySearch);

  var _on2 = _interopRequireDefault(_on);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  var _get = function get(object, property, receiver) {
    if (object === null) object = Function.prototype;
    var desc = Object.getOwnPropertyDescriptor(object, property);

    if (desc === undefined) {
      var parent = Object.getPrototypeOf(object);

      if (parent === null) {
        return undefined;
      } else {
        return get(parent, property, receiver);
      }
    } else if ("value" in desc) {
      return desc.value;
    } else {
      var getter = desc.get;

      if (getter === undefined) {
        return undefined;
      }

      return getter.call(receiver);
    }
  };

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var DetailPageHeader = function (_mixin) {
    _inherits(DetailPageHeader, _mixin);

    /**
     * The Detail Page Header.
     * @extends CreateComponent
     * @extends InitComponentBySearch
     * @param {HTMLElement} element The element working as a page header.
     * @param {Object} [options] The component options.
     */
    function DetailPageHeader(element, options) {
      _classCallCheck(this, DetailPageHeader);

      var _this = _possibleConstructorReturn(this, (DetailPageHeader.__proto__ || Object.getPrototypeOf(DetailPageHeader)).call(this, element, options));

      _this.previousScrollY = 0;
      // Debounce scroll event calls to handleScroll (default: 50)
      var debouncedScroll = (0, _lodash2.default)(_this._handleScroll.bind(_this), 25);
      _this.hScroll = (0, _on2.default)(_this.element.ownerDocument.defaultView, 'scroll', debouncedScroll);
      return _this;
    }

    /**
     * Adds class to header based on users position on the page
     */


    _createClass(DetailPageHeader, [{
      key: '_handleScroll',
      value: function _handleScroll() {
        var scrollPosition = void 0;
        if (this.element.ownerDocument.defaultView.pageYOffset) {
          scrollPosition = this.element.ownerDocument.defaultView.pageYOffset;
        } else {
          scrollPosition = this.element.ownerDocument.defaultView.pageYOffset;
        }

        if (scrollPosition > 86) {
          this.element.dataset.headerActive = true;
          if (scrollPosition < this.previousScrollY) {
            this.element.classList.remove(this.options.scroll);
          } else {
            this.element.classList.add(this.options.scroll);
          }
        } else {
          this.element.classList.remove(this.options.scroll);
          this.element.dataset.headerActive = false;
        }

        this.previousScrollY = scrollPosition;
      }
    }, {
      key: 'release',
      value: function release() {
        this.hScroll.release();
        _get(DetailPageHeader.prototype.__proto__ || Object.getPrototypeOf(DetailPageHeader.prototype), 'release', this).call(this);
      }
    }]);

    return DetailPageHeader;
  }((0, _mixin3.default)(_createComponent2.default, _initComponentBySearch2.default));

  DetailPageHeader.components = new WeakMap();
  DetailPageHeader.options = {
    selectorInit: '[data-detail-page-header]',
    scroll: 'bx--detail-page-header--scroll'
  };
  exports.default = DetailPageHeader;
});
