// carbon-components/umd/components/modal/modal.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', '../../globals/js/misc/mixin', '../../globals/js/mixins/create-component', '../../globals/js/mixins/init-component-by-launcher', '../../globals/js/mixins/evented-show-hide-state', '../../globals/js/misc/event-matches', '../../globals/js/misc/on'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('../../globals/js/misc/mixin'), require('../../globals/js/mixins/create-component'), require('../../globals/js/mixins/init-component-by-launcher'), require('../../globals/js/mixins/evented-show-hide-state'), require('../../globals/js/misc/event-matches'), require('../../globals/js/misc/on'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.mixin, global.createComponent, global.initComponentByLauncher, global.eventedShowHideState, global.eventMatches, global.on);
    global.modal = mod.exports;
  }
})(this, function (exports, _mixin2, _createComponent, _initComponentByLauncher, _eventedShowHideState, _eventMatches, _on) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _mixin3 = _interopRequireDefault(_mixin2);

  var _createComponent2 = _interopRequireDefault(_createComponent);

  var _initComponentByLauncher2 = _interopRequireDefault(_initComponentByLauncher);

  var _eventedShowHideState2 = _interopRequireDefault(_eventedShowHideState);

  var _eventMatches2 = _interopRequireDefault(_eventMatches);

  var _on2 = _interopRequireDefault(_on);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  var _get = function get(object, property, receiver) {
    if (object === null) object = Function.prototype;
    var desc = Object.getOwnPropertyDescriptor(object, property);

    if (desc === undefined) {
      var parent = Object.getPrototypeOf(object);

      if (parent === null) {
        return undefined;
      } else {
        return get(parent, property, receiver);
      }
    } else if ("value" in desc) {
      return desc.value;
    } else {
      var getter = desc.get;

      if (getter === undefined) {
        return undefined;
      }

      return getter.call(receiver);
    }
  };

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var Modal = function (_mixin) {
    _inherits(Modal, _mixin);

    /**
     * Modal dialog.
     * @extends CreateComponent
     * @extends InitComponentByLauncher
     * @extends EventedShowHideState
     * @param {HTMLElement} element The element working as a modal dialog.
     * @param {Object} [options] The component options.
     * @param {string} [options.classVisible] The CSS class for the visible state.
     * @param {string} [options.eventBeforeShown]
     *   The name of the custom event fired before this modal is shown.
     *   Cancellation of this event stops showing the modal.
     * @param {string} [options.eventAfterShown]
     *   The name of the custom event telling that modal is sure shown
     *   without being canceled by the event handler named by `eventBeforeShown` option (`modal-beingshown`).
     * @param {string} [options.eventBeforeHidden]
     *   The name of the custom event fired before this modal is hidden.
     *   Cancellation of this event stops hiding the modal.
     * @param {string} [options.eventAfterHidden]
     *   The name of the custom event telling that modal is sure hidden
     *   without being canceled by the event handler named by `eventBeforeHidden` option (`modal-beinghidden`).
     */
    function Modal(element, options) {
      _classCallCheck(this, Modal);

      var _this = _possibleConstructorReturn(this, (Modal.__proto__ || Object.getPrototypeOf(Modal)).call(this, element, options));

      _this._handleFocusin = function (evt) {
        if (_this.element.classList.contains(_this.options.classVisible) && !_this.element.contains(evt.target)) {
          _this.element.focus();
        }
      };

      _this._hookCloseActions();
      return _this;
    }

    /**
     * A method that runs when `.init()` is called from `initComponentByLauncher`.
     * @param {Event} evt The event fired on the launcher button.
     */


    _createClass(Modal, [{
      key: 'createdByLauncher',
      value: function createdByLauncher(evt) {
        this.show(evt);
      }
    }, {
      key: 'shouldStateBeChanged',
      value: function shouldStateBeChanged(state) {
        if (state === 'shown') {
          return !this.element.classList.contains(this.options.classVisible);
        }

        return this.element.classList.contains(this.options.classVisible);
      }
    }, {
      key: '_changeState',
      value: function _changeState(state, detail, callback) {
        var _this2 = this;

        var transitionEnd = function transitionEnd() {
          _this2.element.removeEventListener('transitionend', transitionEnd);
          if (state === 'shown' && _this2.element.offsetWidth > 0 && _this2.element.offsetHeight > 0) {
            (_this2.element.querySelector(_this2.options.selectorPrimaryFocus) || _this2.element).focus();
          }
          callback();
        };

        if (this._handleFocusinListener) {
          this._handleFocusinListener = this._handleFocusinListener.release();
        }

        if (state === 'shown') {
          var hasFocusin = 'onfocusin' in this.element.ownerDocument.defaultView;
          var focusinEventName = hasFocusin ? 'focusin' : 'focus';
          this._handleFocusinListener = (0, _on2.default)(this.element.ownerDocument, focusinEventName, this._handleFocusin, !hasFocusin);
        }

        if (state === 'hidden') {
          this.element.classList.toggle(this.options.classVisible, false);
        } else if (state === 'shown') {
          this.element.classList.toggle(this.options.classVisible, true);
        }
        this.element.addEventListener('transitionend', transitionEnd);
      }
    }, {
      key: '_hookCloseActions',
      value: function _hookCloseActions() {
        var _this3 = this;

        this.element.addEventListener('click', function (evt) {
          var closeButton = (0, _eventMatches2.default)(evt, _this3.options.selectorModalClose);
          if (closeButton) {
            evt.delegateTarget = closeButton; // eslint-disable-line no-param-reassign
          }
          if (closeButton || evt.target === _this3.element) {
            _this3.hide(evt);
          }
        });

        if (this.keydownHandler) {
          this.element.ownerDocument.body.removeEventListener('keydown', this.keydownHandler);
          this.keydownHandler = null;
        }

        this.keydownHandler = function (evt) {
          if (evt.which === 27) {
            _this3.hide(evt);
          }
        };

        this.element.ownerDocument.body.addEventListener('keydown', this.keydownHandler);
      }
    }, {
      key: 'release',
      value: function release() {
        if (this.keydownHandler) {
          this.element.ownerDocument.body.removeEventListener('keydown', this.keydownHandler);
          this.keydownHandler = null;
        }
        if (this._handleFocusinListener) {
          this._handleFocusinListener = this._handleFocusinListener.release();
        }
        _get(Modal.prototype.__proto__ || Object.getPrototypeOf(Modal.prototype), 'release', this).call(this);
      }
    }]);

    return Modal;
  }((0, _mixin3.default)(_createComponent2.default, _initComponentByLauncher2.default, _eventedShowHideState2.default));

  Modal.components = new WeakMap();
  Modal.options = {
    selectorInit: '[data-modal]',
    selectorModalClose: '[data-modal-close]',
    selectorPrimaryFocus: '[data-modal-primary-focus]',
    classVisible: 'is-visible',
    attribInitTarget: 'data-modal-target',
    initEventNames: ['click'],
    eventBeforeShown: 'modal-beingshown',
    eventAfterShown: 'modal-shown',
    eventBeforeHidden: 'modal-beinghidden',
    eventAfterHidden: 'modal-hidden'
  };
  exports.default = Modal;
});
