// carbon-components/umd/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './globals/js/watch', './components/checkbox/checkbox', './components/file-uploader/file-uploader', './components/fab/fab', './components/content-switcher/content-switcher', './components/tabs/tabs', './components/overflow-menu/overflow-menu', './components/modal/modal', './components/loading/loading', './components/dropdown/dropdown', './components/card/card', './components/number-input/number-input', './components/data-table/data-table', './components/date-picker/date-picker', './components/detail-page-header/detail-page-header', './components/unified-header/left-nav', './components/interior-left-nav/interior-left-nav', './components/unified-header/profile-switcher', './components/pagination/pagination', './components/search/search', './components/accordion/accordion', './components/copy-button/copy-button', './components/notification/notification', './components/toolbar/toolbar', './components/tooltip/tooltip', './components/progress-indicator/progress-indicator', './components/floating-menu/floating-menu', './components/structured-list/structured-list', './components/slider/slider', './components/tile/tile', './components/carousel/carousel', './components/lightbox/lightbox'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./globals/js/watch'), require('./components/checkbox/checkbox'), require('./components/file-uploader/file-uploader'), require('./components/fab/fab'), require('./components/content-switcher/content-switcher'), require('./components/tabs/tabs'), require('./components/overflow-menu/overflow-menu'), require('./components/modal/modal'), require('./components/loading/loading'), require('./components/dropdown/dropdown'), require('./components/card/card'), require('./components/number-input/number-input'), require('./components/data-table/data-table'), require('./components/date-picker/date-picker'), require('./components/detail-page-header/detail-page-header'), require('./components/unified-header/left-nav'), require('./components/interior-left-nav/interior-left-nav'), require('./components/unified-header/profile-switcher'), require('./components/pagination/pagination'), require('./components/search/search'), require('./components/accordion/accordion'), require('./components/copy-button/copy-button'), require('./components/notification/notification'), require('./components/toolbar/toolbar'), require('./components/tooltip/tooltip'), require('./components/progress-indicator/progress-indicator'), require('./components/floating-menu/floating-menu'), require('./components/structured-list/structured-list'), require('./components/slider/slider'), require('./components/tile/tile'), require('./components/carousel/carousel'), require('./components/lightbox/lightbox'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.watch, global.checkbox, global.fileUploader, global.fab, global.contentSwitcher, global.tabs, global.overflowMenu, global.modal, global.loading, global.dropdown, global.card, global.numberInput, global.dataTable, global.datePicker, global.detailPageHeader, global.leftNav, global.interiorLeftNav, global.profileSwitcher, global.pagination, global.search, global.accordion, global.copyButton, global.notification, global.toolbar, global.tooltip, global.progressIndicator, global.floatingMenu, global.structuredList, global.slider, global.tile, global.carousel, global.lightbox);
    global.index = mod.exports;
  }
})(this, function (exports, _watch, _checkbox, _fileUploader, _fab, _contentSwitcher, _tabs, _overflowMenu, _modal, _loading, _dropdown, _card, _numberInput, _dataTable, _datePicker, _detailPageHeader, _leftNav, _interiorLeftNav, _profileSwitcher, _pagination, _search, _accordion, _copyButton, _notification, _toolbar, _tooltip, _progressIndicator, _floatingMenu, _structuredList, _slider, _tile, _carousel, _lightbox) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.componentClasses = exports.Lightbox = exports.Carousel = exports.Tile = exports.Slider = exports.DatePicker = exports.StructuredList = exports.CopyButton = exports.FloatingMenu = exports.ProgressIndicator = exports.Tooltip = exports.Toolbar = exports.Notification = exports.InteriorLeftNav = exports.Accordion = exports.Pagination = exports.ProfileSwitcher = exports.Search = exports.DetailPageHeader = exports.DataTable = exports.NumberInput = exports.Card = exports.Dropdown = exports.Loading = exports.LeftNav = exports.Modal = exports.OverflowMenu = exports.Tab = exports.ContentSwitcher = exports.FileUploader = exports.FabButton = exports.initCheckbox = exports.settings = exports.watch = undefined;
  Object.defineProperty(exports, 'watch', {
    enumerable: true,
    get: function () {
      return _interopRequireDefault(_watch).default;
    }
  });

  var _checkbox2 = _interopRequireDefault(_checkbox);

  var _fileUploader2 = _interopRequireDefault(_fileUploader);

  var _fab2 = _interopRequireDefault(_fab);

  var _contentSwitcher2 = _interopRequireDefault(_contentSwitcher);

  var _tabs2 = _interopRequireDefault(_tabs);

  var _overflowMenu2 = _interopRequireDefault(_overflowMenu);

  var _modal2 = _interopRequireDefault(_modal);

  var _loading2 = _interopRequireDefault(_loading);

  var _dropdown2 = _interopRequireDefault(_dropdown);

  var _card2 = _interopRequireDefault(_card);

  var _numberInput2 = _interopRequireDefault(_numberInput);

  var _dataTable2 = _interopRequireDefault(_dataTable);

  var _datePicker2 = _interopRequireDefault(_datePicker);

  var _detailPageHeader2 = _interopRequireDefault(_detailPageHeader);

  var _leftNav2 = _interopRequireDefault(_leftNav);

  var _interiorLeftNav2 = _interopRequireDefault(_interiorLeftNav);

  var _profileSwitcher2 = _interopRequireDefault(_profileSwitcher);

  var _pagination2 = _interopRequireDefault(_pagination);

  var _search2 = _interopRequireDefault(_search);

  var _accordion2 = _interopRequireDefault(_accordion);

  var _copyButton2 = _interopRequireDefault(_copyButton);

  var _notification2 = _interopRequireDefault(_notification);

  var _toolbar2 = _interopRequireDefault(_toolbar);

  var _tooltip2 = _interopRequireDefault(_tooltip);

  var _progressIndicator2 = _interopRequireDefault(_progressIndicator);

  var _floatingMenu2 = _interopRequireDefault(_floatingMenu);

  var _structuredList2 = _interopRequireDefault(_structuredList);

  var _slider2 = _interopRequireDefault(_slider);

  var _tile2 = _interopRequireDefault(_tile);

  var _carousel2 = _interopRequireDefault(_carousel);

  var _lightbox2 = _interopRequireDefault(_lightbox);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var settings = {};

  /**
   * This module is used for the following purposes:
   * 1. Export ES2015 classes as modules (used with base-elements and components)
   * 2. Build an ES5-compatible files for prototyping.
   *    See /path/to/bluemix-components/dist/dist-demo.html for details.
   * @exports CarbomComponents
   * @example <caption>Consume ES2015 modules from this file using import (Usage pattern 1.)</caption>
   * import { Fab, FileUploader } from '/path/to/your/project/node_modules/@console/bluemix-components';
   */
  exports.settings = settings;
  exports.initCheckbox = _checkbox2.default;
  exports.FabButton = _fab2.default;
  exports.FileUploader = _fileUploader2.default;
  exports.ContentSwitcher = _contentSwitcher2.default;
  exports.Tab = _tabs2.default;
  exports.OverflowMenu = _overflowMenu2.default;
  exports.Modal = _modal2.default;
  exports.LeftNav = _leftNav2.default;
  exports.Loading = _loading2.default;
  exports.Dropdown = _dropdown2.default;
  exports.Card = _card2.default;
  exports.NumberInput = _numberInput2.default;
  exports.DataTable = _dataTable2.default;
  exports.DetailPageHeader = _detailPageHeader2.default;
  exports.Search = _search2.default;
  exports.ProfileSwitcher = _profileSwitcher2.default;
  exports.Pagination = _pagination2.default;
  exports.Accordion = _accordion2.default;
  exports.InteriorLeftNav = _interiorLeftNav2.default;
  exports.Notification = _notification2.default;
  exports.Toolbar = _toolbar2.default;
  exports.Tooltip = _tooltip2.default;
  exports.ProgressIndicator = _progressIndicator2.default;
  exports.FloatingMenu = _floatingMenu2.default;
  exports.CopyButton = _copyButton2.default;
  exports.StructuredList = _structuredList2.default;
  exports.DatePicker = _datePicker2.default;
  exports.Slider = _slider2.default;
  exports.Tile = _tile2.default;
  exports.Carousel = _carousel2.default;
  exports.Lightbox = _lightbox2.default;


  /**
   * List of component classes to be auto-instantiated.
   * @private
   */
  var componentClasses = exports.componentClasses = [_fab2.default, _fileUploader2.default, _contentSwitcher2.default, _tabs2.default, _overflowMenu2.default, _modal2.default, _loading2.default, _dropdown2.default, _card2.default, _numberInput2.default, _dataTable2.default, _detailPageHeader2.default, _leftNav2.default, _interiorLeftNav2.default, _profileSwitcher2.default, _pagination2.default, _search2.default, _accordion2.default, _copyButton2.default, _notification2.default, _toolbar2.default, _tooltip2.default, _progressIndicator2.default, _structuredList2.default, _datePicker2.default, _slider2.default, _tile2.default, _carousel2.default, _lightbox2.default];

  /**
   * Instantiates components automatically
   * by searching for elements with `data-component-name` (e.g. `data-loading`) attribute
   * or upon DOM events (e.g. clicking) on such elements.
   * See each components' static `.init()` methods for details.
   * @private
   */
  var init = function init() {
    if (!settings.disableAutoInit) {
      (0, _checkbox2.default)();
      componentClasses.forEach(function (Clz) {
        Clz.init();
      });
    }
  };

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', init);
  } else {
    // DOMContentLoaded has been fired already
    // Let consumer have chance to see if it wants automatic instantiation disabled, and then run automatic instantiation otherwise
    setTimeout(init, 0);
  }
});
